### DISCUSSION TITLE
How to get webpack honour swiperjs module imports, avoiding bundling the ones not used?

### DISCUSSION THREAD

**Discussion by stratboy** [webpack/webpack#15341] [Q&A]
Hi, swiperjs, from docs:

"By default Swiper exports only core version without additional modules (like Navigation, Pagination, etc.). So you need to import and configure them too:"

I'm doing some tests, my index.js is just this:

```js
import Swiper, { Navigation, Pagination } from 'swiper';
console.log(Swiper);
```

And this is my webpack config:

```js
const path = require('path');
const BundleAnalyzerPlugin = require('webpack-bundle-analyzer').BundleAnalyzerPlugin;

module.exports = {
  //mode: 'development',
  mode: 'production',
  watch: true,
  entry: {
    index: './src/index.js',

  },

  output: {
    filename: '[name].js',
    path: path.resolve(__dirname, 'dist'),
    clean: true,
  },


  plugins: [
    new BundleAnalyzerPlugin()
  ],

  module: {
    rules: [
      {
        test: /\.m?js$/,
        exclude: /node_modules/,
        use: {
          loader: "babel-loader",
          options: { presets: ['@babel/preset-env'] }
        }
      } // babel
    ] // rules
  } // module
};
```

Well, the BundleAnalyzer graph shows that ALL the swiper modules are bundled. Why? How can I avoid that?

**Comment by alexander-akait** [comment]
It means swiper import them itself, you can ignore them if you need using https://webpack.js.org/plugins/ignore-plugin/, but it can break code if code did not written in right way

**Reply by stratboy** [reply]
> It means swiper import them itself, you can ignore them if you need using https://webpack.js.org/plugins/ignore-plugin/, but it can break code if code did not written in right way

Hi, I don't know... I tried the same thing with Rollup and it works correctly (but I'd prefer using webpack). Also, it sounds very strange that Swiper declares a feature that doesn't exists, also considering it's there since v7 (now 8.0.3).

Also, I think what's important is not what swiper imports: it's what it exports. At least for what I understood about modules till now.

**Reply by alexander-akait** [reply]
If you need answer, please create reproducible example, I don't know your configuration/code

webpack doesn't import if code doesn't contain `import ... from ...;`

**Reply by stratboy** [reply]
> If you need answer, please create reproducible example, I don't know your configuration/code
> 
> webpack doesn't import if code doesn't contain `import ... from ...;`

Sure, but I already copied my index.js and my webpack configuration: what other code/files do you need?

**Reply by alexander-akait** [reply]
Look at `node_modules/swiper/package.json`:
```
  "exports": {
    ".": "./swiper.esm.js",
```

it means we will bundle `swiper.esm.js`, content of `swiper.esm.js`:
```js
/**
 * Swiper 8.0.3
 * Most modern mobile touch slider and framework with hardware accelerated transitions
 * https://swiperjs.com
 *
 * Copyright 2014-2022 Vladimir Kharlampidi
 *
 * Released under the MIT License
 *
 * Released on: February 3, 2022
 */

export { default as Swiper, default } from './core/core.js';
export { default as Virtual } from './modules/virtual/virtual.js';
export { default as Keyboard } from './modules/keyboard/keyboard.js';
export { default as Mousewheel } from './modules/mousewheel/mousewheel.js';
export { default as Navigation } from './modules/navigation/navigation.js';
export { default as Pagination } from './modules/pagination/pagination.js';
export { default as Scrollbar } from './modules/scrollbar/scrollbar.js';
export { default as Parallax } from './modules/parallax/parallax.js';
export { default as Zoom } from './modules/zoom/zoom.js';
export { default as Lazy } from './modules/lazy/lazy.js';
export { default as Controller } from './modules/controller/controller.js';
export { default as A11y } from './modules/a11y/a11y.js';
export { default as History } from './modules/history/history.js';
export { default as HashNavigation } from './modules/hash-navigation/hash-navigation.js';
export { default as Autoplay } from './modules/autoplay/autoplay.js';
export { default as Thumbs } from './modules/thumbs/thumbs.js';
export { default as FreeMode } from './modules/free-mode/free-mode.js';
export { default as Grid } from './modules/grid/grid.js';
export { default as Manipulation } from './modules/manipulation/manipulation.js';
export { default as EffectFade } from './modules/effect-fade/effect-fade.js';
export { default as EffectCube } from './modules/effect-cube/effect-cube.js';
export { default as EffectFlip } from './modules/effect-flip/effect-flip.js';
export { default as EffectCoverflow } from './modules/effect-coverflow/effect-coverflow.js';
export { default as EffectCreative } from './modules/effect-creative/effect-creative.js';
export { default as EffectCards } from './modules/effect-cards/effect-cards.js';
```

**Reply by alexander-akait** [reply]
Swiper need to set https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free, i.e. put ` "sideEffects": false` in `package.json`

**Reply by alexander-akait** [reply]
Or you can set `sideEffects` on loader level https://webpack.js.org/configuration/module/#rulesideeffects, just create loader with regexp to swipper and set `sideEffects: false`

**Reply by stratboy** [reply]
> Or you can set `sideEffects` on loader level https://webpack.js.org/configuration/module/#rulesideeffects, just create loader with regexp to swipper and set `sideEffects: false`

I already did in package.json, but didn't work. Now I'll try on the loader and let you know.

**Reply by stratboy** [reply]
Yes! sideEffects at module level does work!

```js
  module: {
    rules: [
      // [...] here normally there are babel settings and the like
      
      { // here doing the swiper loader and declaring no sideEffects
        test: /swiper\.esm\.js/,
        sideEffects: false
      }
    ] // rules
  } // module
```

Why didn't it worked when set in package.json?

**Reply by alexander-akait** [reply]
Because you need mark all swipes sources, just mark all sources in `swiper` as sideEffect free

**Reply by alexander-akait** [reply]
> I already did in package.json, but didn't work. Now I'll try on the loader and let you know.

Note - webpack drop them only when you remove it from `import ... from ...;` (i.e. unused export), if you have Navigation in import, webpack keeps them

**Reply by stratboy** [reply]
> Because you need mark all swipes sources, just mark all sources in `swiper` as sideEffect free

The loader version worked fine. Shouldn't the package.json version be 'global'? One for all without the need to specify at loader level?

**Reply by alexander-akait** [reply]
But swiper don't have `sideEffects: false` in `package.json`

**Reply by alexander-akait** [reply]
https://github.com/nolimits4web/swiper/blob/master/package.json
